### PR TITLE
Fix truncate

### PR DESCRIPTION
**Pull request by lgvz**
Truncation didn't work for similar reasons as writing.

Another problem was that `prepare_window` didn't make sure sequencer had enough room for the content of the last chunk to be decrypted. And sure enough, this resulted in index-out-of-bounds when initializing `SelfEncryptor` with a `DataMap::Chunks` - which leaves sequencer's length to zero - and then truncating.

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/self_encryption/159)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @maqi

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: src/lib.rs [hunk 1]
```diff
@@ -564,26 +564,33 @@ impl<S: Storage + Send + Sync + 'static> SelfEncryptor<S> {
...
-        if position < old_size {
-            self.sequencer.truncate(position as usize);
-            let last_chunk = self.get_chunk_number(position) + 1;
-            self.chunks.truncate(last_chunk as usize);
+        if new_size < old_size {
+            self.prepare_window(new_size, 0);   // call before changing self.file_size
+            self.file_size = new_size;
+            self.sequencer.truncate(new_size as usize);
+            let num_chunks = self.get_num_chunks(); // call with updated self.file_size
+            self.chunks.truncate(num_chunks as usize);
```
Comment (lgvz) [inline src/lib.rs]: This is flaky because `self.chunks` is not guaranteed to be sorted. This lib has a lot of problems...

Comment (Fraser999) [inline src/lib.rs]: @lgvz Agreed.  Would it be possible for you to contact me directly to discuss this further?  fraser.hutchison@maidsafe.net

**Comment by lgvz** [comment]
It's me: 34gv...ogw1

**Comment by dirvine** [comment]
:+1:  All the info we need for this stuff, so cool and thx again. @NickLambert will sort out some btc as a thanks from us and the community you are helping here.

**Comment by lgvz** [comment]
Thank you! :)

**Comment by NickLambert** [comment]
Thanks @lgvz, 0.4 BTC winging its way to you.

**Pull request merged**